### PR TITLE
Implement stochastic gradient descent

### PR DESCRIPTION
**Pull request by rouille**
### Purpose
Implement batch stochastic gradient descent. The algorithm identifies the parameters of a model that correspond to the best fit fit between predicted and actual outputs.

### What the code is doing
It is an iterative process. you start with an arbitrarily chosen value for the parameters, update each parameter to new value in the direction of their negative gradient $v --> v - \eta \Delta C$, where $\eta$ is the learning rate, a positive parameter controlling the size of the update, and $C$ is the function we we want to minimize. The implemented algorithm also uses a decay rate to adapt the learning rate during the algorithm execution by taking into account the magnitude and direction of the previous update. In this implementation the dataset in divided in batches and the gradient is updated for each batch. 

### Testing
New unit test.

Generate $n$ pairs ($X$, $y$) related via $y = X  \beta + \epsilon$, where:
* $\beta = {\left\lbrack \matrix{1 \cr 1 \cr 10} \right\rbrack}$ is the parameter vector;
* $X = {\left\lbrack \matrix{1 & x_{11} & x_{12} \cr \vdots & \ddots & \vdots \cr 1 & x_{n1} & x_{n2}} \right\rbrack}$ is the regressor, where $x_{ij}$ are randomly drawn from a normal distribution;
* $\epsilon = {\left\lbrack \matrix{\epsilon_1 \cr \vdots \cr \epsilon_n} \right\rbrack}$ is an error variable that adds noise to the linear relationship between the dependent variable $y$ and regressor $X$. the elements are randomly drawn from a normal distribution.

The sum of squared residuals is used as the cost function and the parameters ($\beta$) are inferred using stochastic gradient descent. The value of the parameters are then compared to the estimates from an ordinary least squares 

### Where to look
The `interview.algorithm.sgd` encloses the `SGD` class

### Usage Example/Visuals
N/A

### Time estimate
20min

### CODE REVIEW

File: interview/algorithm/sgd.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        batch_size=1,
+        tolerance=1e-3,
+        seed=None,
+    ):
+        self.learning_rate = learning_rate
+        self.decay_rate = decay_rate
+        self.max_iteration = max_iteration
+        self.batch_size = batch_size
+        self.tolerance = tolerance
+        self.seed = None if seed is None else seed
```
Comment (kasparm) [inline interview/algorithm/sgd.py]: Why are you doing this? Is self.seed not given by the input seed?

File: interview/algorithm/sgd.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                stop = start + self.batch_size
+                x_batch, y_batch = x[start:stop], y[start:stop]
+                grad = np.array(gradient(x_batch, y_batch, var))
+                diff = self.decay_rate * diff - self.learning_rate * grad
+
+                if np.linalg.norm(grad) < self.tolerance:
+                    break
+
+                var += diff
+        return var
```
Comment (BainanXia) [inline interview/algorithm/sgd.py]: In practice we might want to know how the fit ends, i.e. maximum iteration reached or tolerance met.

File: interview/algorithm/sgd.py (hunk 1 above)
Comment (rouille) [reply interview/algorithm/sgd.py]: Not sure what I was trying to do here.

File: interview/algorithm/sgd.py (hunk 2 above)
Comment (rouille) [reply interview/algorithm/sgd.py]: Good point!

File: interview/algorithm/sgd.py (hunk 1 above)
Comment (rouille) [reply interview/algorithm/sgd.py]: Done

File: interview/algorithm/sgd.py (hunk 2 above)
Comment (rouille) [reply interview/algorithm/sgd.py]: I have slightly refactored the code in the last commit ([c6196a5](https://github.com/rouille/Interview/pull/8/commits/c6196a5dc975c2b841c9437cc741d7461f0ab3aa)). I have added 2 methods to the `SGD` class: `get_params` that returns the parameters that minimize the cost function and `print_fit_info` that prints information on the convergence.
